File: src/app/api/conversations.py
import os
from pathlib import Path
from typing import cast

from fastapi import APIRouter, HTTPException, Response, status

from app.api.schemas import (
    ConversationAppendEventsRequest,
    ConversationAppendEventsResponse,
    ConversationRewriteRequest,
    ConversationRewriteResponse,
    ConversationContextStatusView,
    ConversationEventsPageView,
    ConversationTokenUsageView,
    ConversationCreateRequest,
    ConversationCreateResponse,
    ConversationDetailView,
    ConversationSummaryView,
)
from app.db.repositories.model_usage import sum_conversation_usage
from app.db.session import Session, engine
from app.services.context_manager import ContextManager, JsonObject
from app.services.conversation_service import ConversationService
from app.services.model_service import ModelService

router = APIRouter()


def get_conversation_service() -> ConversationService:
    configured = os.getenv("OPS_AGENT_CONVERSATIONS_DIR", "")
    base_dir = Path(configured) if configured else Path.cwd() / ".ops-agent" / "conversations"
    return ConversationService(base_dir=base_dir, model_service=ModelService())


@router.get("/api/conversations", response_model=list[ConversationSummaryView])
def list_conversations() -> list[ConversationSummaryView]:
    service = get_conversation_service()
    return [ConversationSummaryView.model_validate(summary.__dict__) for summary in service.list_conversations()]


@router.post("/api/conversations", response_model=ConversationCreateResponse)
def create_conversation(payload: ConversationCreateRequest) -> ConversationCreateResponse:
    service = get_conversation_service()
    summary = service.create_conversation(
        selected_model=payload.selected_model,
        asset_id=payload.asset_id,
        scope_mode=payload.scope_mode,
    )
    return ConversationCreateResponse(
        conversation=ConversationSummaryView.model_validate(summary.__dict__),
        events=[],
    )


@router.get("/api/conversations/{conversation_id}", response_model=ConversationDetailView)
def get_conversation(conversation_id: str) -> ConversationDetailView:
    service = get_conversation_service()
    try:
        detail = service.get_conversation(conversation_id)
    except FileNotFoundError as exc:
        raise HTTPException(status_code=404, detail="Conversation not found") from exc
    return ConversationDetailView.model_validate(detail.__dict__)


@router.get("/api/conversations/{conversation_id}/events", response_model=ConversationEventsPageView)
def get_conversation_events(conversation_id: str, offset: int = 0, limit: int = 200, tail: int | None = None) -> ConversationEventsPageView:
    service = get_conversation_service()
    try:
        page = service.get_events_tail(conversation_id, limit=tail) if tail is not None else service.get_events_page(conversation_id, offset=offset, limit=limit)
    except FileNotFoundError as exc:
        raise HTTPException(status_code=404, detail="Conversation not found") from exc
    return ConversationEventsPageView(
        conversation=ConversationSummaryView.model_validate(page.conversation.__dict__),
        events=page.events,
        offset=page.offset,
        limit=page.limit,
        total=page.total,
        has_more_before=page.has_more_before,
        has_more_after=page.has_more_after,
    )


@router.post("/api/conversations/{conversation_id}/events", response_model=ConversationAppendEventsResponse)
def append_conversation_events(conversation_id: str, payload: ConversationAppendEventsRequest) -> ConversationAppendEventsResponse:
    service = get_conversation_service()
    try:
        detail = service.append_events(conversation_id, payload.events)
    except FileNotFoundError as exc:
        raise HTTPException(status_code=404, detail="Conversation not found") from exc
    return ConversationAppendEventsResponse(
        conversation=ConversationSummaryView.model_validate(service.to_summary(detail).__dict__),
        appended_count=len(payload.events),
    )


@router.get("/api/conversations/{conversation_id}/context", response_model=ConversationContextStatusView)
def get_conversation_context(conversation_id: str) -> ConversationContextStatusView:
    service = get_conversation_service()
    try:
        detail = service.get_conversation(conversation_id)
    except FileNotFoundError as exc:
        raise HTTPException(status_code=404, detail="Conversation not found") from exc

    context_manager = ContextManager(service.base_dir / "context")
    events = cast(list[JsonObject], detail.events or [])
    with Session(engine) as session:
        usage = sum_conversation_usage(session, conversation_id)
    token_usage = ConversationTokenUsageView(
        input_tokens=usage.input_tokens,
        output_tokens=usage.output_tokens,
        cache_creation_input_tokens=usage.cache_creation_input_tokens,
        cache_read_input_tokens=usage.cache_read_input_tokens,
        total_tokens=usage.total_tokens,
        measurement="reported" if usage.total_tokens > 0 else "unavailable",
    )
    metadata = context_manager.read_metadata(conversation_id)
    source_revision = context_manager.source_revision(events)
    if metadata is None or metadata.source_conversation_revision != source_revision:
        model_config = ModelService().load_settings()
        result = context_manager.prepare_context(conversation_id, events, model_config)
        return ConversationContextStatusView(
            context_percent=result.context_percent,
            context_status=result.context_status,
            token_usage=token_usage,
        )
    return ConversationContextStatusView(
        context_percent=metadata.context_percent,
        context_status=metadata.context_status,
        token_usage=token_usage,
    )


@router.post("/api/conversations/{conversation_id}/rewrite", response_model=ConversationRewriteResponse)
def rewrite_conversation(conversation_id: str, payload: ConversationRewriteRequest) -> ConversationRewriteResponse:
    from app.api.console import get_console_app_service

    active_runtimes = [
        runtime
        for runtime in get_console_app_service().runtime_manager.list_runtimes(conversation_id)
        if not runtime.state.is_terminal()
    ]
    if active_runtimes:
        raise HTTPException(status_code=409, detail="当前 Agent 任务尚未结束，不能改写会话历史。")
    service = get_conversation_service()
    try:
        detail = service.truncate_before_event(conversation_id, payload.before_event_id)
    except FileNotFoundError as exc:
        raise HTTPException(status_code=404, detail="Conversation not found") from exc
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    return ConversationRewriteResponse(
        conversation=ConversationSummaryView.model_validate(service.to_summary(detail).__dict__),
        events=detail.events,
    )


@router.delete("/api/conversations/{conversation_id}", status_code=status.HTTP_204_NO_CONTENT, response_class=Response)
def delete_conversation(conversation_id: str, cancel_active: bool = False) -> Response:
    service = get_conversation_service()
    try:
        service.get_conversation(conversation_id)
    except FileNotFoundError as exc:
        raise HTTPException(status_code=404, detail="Conversation not found") from exc

    # Imported lazily because console routes also use get_conversation_service.
    from app.api.console import get_console_app_service

    runtime_manager = get_console_app_service().runtime_manager
    active_runtimes = [runtime for runtime in runtime_manager.list_runtimes(conversation_id) if not runtime.state.is_terminal()]
    if active_runtimes and not cancel_active:
        raise HTTPException(
            status_code=409,
            detail="该会话仍有未结束的 Agent 任务。确认停止任务后才能删除。",
        )
    for runtime in active_runtimes:
        runtime_manager.cancel(runtime.runtime_id, reason="Conversation deleted by operator.")

    remaining_active = [runtime for runtime in runtime_manager.list_runtimes(conversation_id) if not runtime.state.is_terminal()]
    if remaining_active:
        raise HTTPException(
            status_code=409,
            detail="已请求停止 Agent 任务；任务完全停止后请再次删除。",
        )

    try:
        service.delete_conversation(conversation_id)
    except FileNotFoundError as exc:
        raise HTTPException(status_code=404, detail="Conversation not found") from exc
    runtime_manager.forget_conversation_runtimes(conversation_id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)

File: web/src/api/conversations.ts
import { requestJson, requestVoid } from './client'
import { mapRequiredTimestamps } from './mappers'
import type { ConversationAppendEventsResponseDto, ConversationContextStatusDto, ConversationCreateResponseDto, ConversationDetailDto, ConversationEventsPageDto, ConversationSummaryDto } from '../types/api'
import type { ConversationContextStatus, ConversationDetail, ConversationEventsPage, ConversationSummary, EventItem } from '../types/ops'

export function mapConversationSummary(dto: ConversationSummaryDto): ConversationSummary {
  return {
    id: dto.id,
    title: dto.title,
    selectedModel: dto.selected_model,
    ...mapRequiredTimestamps(dto),
    eventCount: dto.event_count,
    lastEventKind: dto.last_event_kind,
    assetId: dto.asset_id,
    scopeMode: dto.scope_mode,
    allowedAssetIds: dto.allowed_asset_ids,
  }
}

export function mapConversationDetail(dto: ConversationDetailDto): ConversationDetail {
  return {
    id: dto.id,
    title: dto.title,
    selectedModel: dto.selected_model,
    ...mapRequiredTimestamps(dto),
    events: dto.events,
    assetId: dto.asset_id,
    scopeMode: dto.scope_mode,
    allowedAssetIds: dto.allowed_asset_ids,
  }
}

export function mapConversationEventsPage(dto: ConversationEventsPageDto): ConversationEventsPage {
  return {
    conversation: mapConversationSummary(dto.conversation),
    events: dto.events,
    offset: dto.offset,
    limit: dto.limit,
    total: dto.total,
    hasMoreBefore: dto.has_more_before,
    hasMoreAfter: dto.has_more_after,
  }
}

export function mapConversationAppendResponse(dto: ConversationAppendEventsResponseDto): { conversation: ConversationSummary; appendedCount: number } {
  return {
    conversation: mapConversationSummary(dto.conversation),
    appendedCount: dto.appended_count,
  }
}

export function mapConversationContextStatus(dto: ConversationContextStatusDto): ConversationContextStatus {
  return {
    contextPercent: dto.context_percent,
    contextStatus: dto.context_status,
    tokenUsage: dto.token_usage
      ? {
          inputTokens: dto.token_usage.input_tokens,
          outputTokens: dto.token_usage.output_tokens,
          cacheCreationInputTokens: dto.token_usage.cache_creation_input_tokens,
          cacheReadInputTokens: dto.token_usage.cache_read_input_tokens,
          totalTokens: dto.token_usage.total_tokens,
          measurement: dto.token_usage.measurement,
        }
      : undefined,
  }
}

export async function getConversations(): Promise<ConversationSummary[]> {
  const conversations = await requestJson<ConversationSummaryDto[]>('/api/conversations')
  return conversations.map(mapConversationSummary)
}

export async function createConversation(
  selectedModel: string | null,
  assetId: number,
  scopeMode: 'single' | 'multi' = 'single',
): Promise<{ conversation: ConversationSummary; events: EventItem[] }> {
  const response = await requestJson<ConversationCreateResponseDto>('/api/conversations', {
    method: 'POST',
    body: JSON.stringify({ selected_model: selectedModel, asset_id: assetId, scope_mode: scopeMode }),
  })
  return {
    conversation: mapConversationSummary(response.conversation),
    events: response.events,
  }
}

export async function rewriteConversation(
  conversationId: string,
  beforeEventId: string,
): Promise<{ conversation: ConversationSummary; events: EventItem[] }> {
  const response = await requestJson<ConversationCreateResponseDto>(`/api/conversations/${conversationId}/rewrite`, {
    method: 'POST',
    body: JSON.stringify({ before_event_id: beforeEventId }),
  })
  return {
    conversation: mapConversationSummary(response.conversation),
    events: response.events,
  }
}

export async function getConversation(conversationId: string): Promise<ConversationDetail> {
  const conversation = await requestJson<ConversationDetailDto>(`/api/conversations/${conversationId}`)
  return mapConversationDetail(conversation)
}

export async function getConversationEventsTail(conversationId: string, limit = 200): Promise<ConversationEventsPage> {
  const page = await requestJson<ConversationEventsPageDto>(`/api/conversations/${conversationId}/events?tail=${limit}`)
  return mapConversationEventsPage(page)
}

export async function getConversationEventsPage(conversationId: string, offset: number, limit = 200): Promise<ConversationEventsPage> {
  const page = await requestJson<ConversationEventsPageDto>(`/api/conversations/${conversationId}/events?offset=${offset}&limit=${limit}`)
  return mapConversationEventsPage(page)
}

export async function getConversationContext(conversationId: string): Promise<ConversationContextStatus> {
  const status = await requestJson<ConversationContextStatusDto>(`/api/conversations/${conversationId}/context`)
  return mapConversationContextStatus(status)
}

export async function appendConversationEvents(conversationId: string, events: EventItem[]): Promise<{ conversation: ConversationSummary; appendedCount: number }> {
  const response = await requestJson<ConversationAppendEventsResponseDto>(`/api/conversations/${conversationId}/events`, {
    method: 'POST',
    body: JSON.stringify({ events }),
  })
  return mapConversationAppendResponse(response)
}

export async function deleteConversation(conversationId: string, cancelActive = false): Promise<void> {
  const query = cancelActive ? '?cancel_active=true' : ''
  await requestVoid(`/api/conversations/${conversationId}${query}`, {
    method: 'DELETE',
  })
}
